### ISSUE TITLE
proposal: gofmt should allow short inlined if-return statements

### ISSUE THREAD

**Issue by eldanelias** [golang/go#27135]
You've seen it everywhere:
```
if err != nil {
    return err
}
```

Let's allow `if err != nil { return err }`

**Comment by ALTree** [comment]
Statements that drastically change the program's control flow (like `return`) should be prominent and easy to spot. Inlining an `if`-`return` like that makes it less visible. It may be ok when the return statement is the whole body of the function, like in

```
func add(a, b int) { return a + b }
```

(which is allowed by `gofmt`), but I don't think it is when it's a jump buried inside an `if` somewhere in the function.

**Comment by eldanelias** [comment]
@ALTree if functions are so big you can't spot a return statement inside an if statement then the issue is with the cleanliness of the code, I think allowing to inline if-return will help reduce the clutter in the eyes.

Small functions or even just functions that adhere to Command & Query separation will not have the issue you describe.

Thanks for the quick reply!

**Comment by bronze1man** [comment]
@FlorianUekermann 

> Adding distance between lines of code makes code more difficult to understand and read. People forget stuff if they need to scan or scroll.

People forget stuff when there is too much information any way. Making the code less line can not solve this problem. 
Two statement one line may need the brain store more information, like there is two thing in that line, one thing in this line, the horizontal position on some statement.  If all lines only have one statement, the brain do not need to store that information, it just need remember the vertical position of the statement.
If someone is too smart, he can understand the code with two statements in one line or one statement in one line easily. But I think he should sympathize his colleague which is not so smart.

> Vertical space is much scarcer than horizontal space nowadays (with most people having 16:9 screens or similar).

The screen is big so you can see two files or three files in horizontal parallel.

**Comment by eldanelias** [comment]
@bronze1man
I agree that generally one line should do 1 thing but in my experience when the line is short enough if-returns as well as terinary operator usages are easy enough to follow, I agree that they could potentially be missed in long functions but for short ones that issue doesn't exist IMHO

It seems they are going to solve the clutter issue in go 2 by having a block of error handling, so that might solve the core issue

**Comment by networkimprov** [comment]
See also...
Go2 drafts error handling: https://go.googlesource.com/proposal/+/master/design/go2draft.md
Community feedback: https://github.com/golang/go/wiki/Go2ErrorHandlingFeedback
